riot/venv.py

```diff
@@ -0,0 +1,206 @@
+import dataclasses
+import logging
+import os.path
+import subprocess
+import typing as t
+
+from riot.interpreter import Interpreter
+from riot.utils import (
+    CmdFailure,
+    expand_specs,
+    get_pep_dep,
+    rm_singletons,
+    rmchars,
+    run_cmd,
+    run_cmd_venv,
+    to_list,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class VenvError(Exception):
+    pass
+
+
+@dataclasses.dataclass
+class Venv:
+    pkgs: t.Tuple[t.Tuple[str, str]]
+    py: t.Optional[Interpreter]
+    parent: t.Optional["Venv"] = None
+
+    @property
+    def venv_path(self) -> t.Optional[str]:
+        """Return path to directory of the venv this instance should use.
+
+        This will return a python version + package specific venv path name.
+
+        If no packages are defined it will return the ``Interpreter.venv_path``.
+        """
+        if self.py is None:
+            return None
+
+        venv_path = self.py.venv_path
+        if self.needs_venv:
+            venv_path = "_".join(
+                [venv_path] + [f"{n}{rmchars('<=>.,', v)}" for n, v in self.pkgs]
+            )
+        return venv_path
+
+    @property
+    def target(self) -> t.Optional[str]:
+        if self.venv_path is None:
+            return None
+        return os.path.join(self.venv_path, "target")
+
+    @property
+    def needs_venv(self) -> bool:
+        """Whether this ``VenvInstance`` needs its own venv or not."""
+        return bool(self.pkg_str)
+
+    @property
+    def pkg_str(self) -> str:
+        """Return pip friendly install string from defined packages."""
+        return " ".join(
+            [
+                f"'{get_pep_dep(lib, version)}'"
+                for lib, version in self.pkgs
+                if version is not None
+            ]
+        )
+
+    @property
+    def pythonpath(self):
```

Comment:
Don't mind the debug noise. By walking up the parenting relation we can determine the PYTHONPATH to use in order to import the correct dependencies 🤞